database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -0,0 +1,172 @@
+            ref.removeEventListener(mRefs.remove(ref));
+        }
+    }
+
+    @Override
+    public int getCount() {
+        return mDataSnapshots.size();
+    }
+
+    @Override
+    public DataSnapshot getItem(int index) {
+        return mDataSnapshots.get(index);
+    }
+
+    private int getIndexForKey(String key) {
+        int dataCount = getCount();
+        int index = 0;
+        for (int keyIndex = 0; index < dataCount; keyIndex++) {
+            String superKey = super.getItem(keyIndex).getKey();
+            if (key.equals(superKey)) {
+                break;
+            } else if (getItem(index).getKey().equals(superKey)) {
+                index++;
+            }
+        }
+        return index;
+    }
+
+    private boolean isMatch(int index, String key) {
+        return index >= 0 && index < getCount() && getItem(index).getKey().equals(key);
+    }
+
+    @Override
+    public void onChildAdded(DataSnapshot keySnapshot, String previousChildKey) {
+        super.setOnChangedListener(null);
+        super.onChildAdded(keySnapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+
+        DatabaseReference ref = mRef.child(keySnapshot.getKey());
+        mRefs.put(ref, ref.addValueEventListener(new DataRefSnapshot()));
+    }
+
+    @Override
+    public void onChildChanged(DataSnapshot snapshot, String previousChildKey) {
+        super.setOnChangedListener(null);
+        super.onChildChanged(snapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+    }
+
+    @Override
+    public void onChildRemoved(DataSnapshot keySnapshot) {
+        String key = keySnapshot.getKey();
+        int index = getIndexForKey(key);
+        mRef.child(key).removeEventListener(mRefs.remove(mRef.child(key)));
+
+        super.setOnChangedListener(null);
+        super.onChildRemoved(keySnapshot);
+        super.setOnChangedListener(mListener);
+
+        if (isMatch(index, key)) {
+            mDataSnapshots.remove(index);
+            notifyChangedListeners(OnChangedListener.EventType.REMOVED, index);
+        }
+    }
+
+    @Override
+    public void onChildMoved(DataSnapshot keySnapshot, String previousChildKey) {
+        String key = keySnapshot.getKey();
+        int oldIndex = getIndexForKey(key);
+
+        super.setOnChangedListener(null);
+        super.onChildMoved(keySnapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+
+        if (isMatch(oldIndex, key)) {
+            DataSnapshot snapshot = mDataSnapshots.remove(oldIndex);
+            int newIndex = getIndexForKey(key);
+            mDataSnapshots.add(newIndex, snapshot);
+            notifyChangedListeners(OnChangedListener.EventType.MOVED, newIndex, oldIndex);
+        }
+    }
+
+    @Override
+    public void onCancelled(DatabaseError error) {
+        Log.e(TAG, "A fatal error occur retrieving the necessary keys to poplulate your adapter");
+        super.onCancelled(error);
+    }
```

Comment:
Devs can override [this method](url) in which they could reinitialize their adapter with a new ref or replace their view with an "Oops, an unexpected error occurred." message. I think this gives devs nice options. What do you think of having a separate error method to override for the key failure vs the ref failure?